### PR TITLE
Nadro/3686/file swapping while executing

### PR DESCRIPTION
**Pull request by nadr0**
closes https://github.com/KittyCAD/modeling-app/issues/3686
related to https://github.com/KittyCAD/modeling-app/issues/3470

**Technical bits**

I implemented a promise rejection on the `pendingCommands` which sends an error to the KCL code then we can catch it within the JS `_executor` function as well as propagating it through the `executeAst` call. 

- After we interrupt the `executeAst` don't run some scene setup since we will be immediately starting a new scene, only run the cleanup. 
- I detect the custom reject error within `langHelper's` `executeAst` to detect a KCL error that isn't actually a KCL error. It is my custom error. This allows me to say the commands will be interrupted. 
- TODO, `sendSceneCommand` had an uncaught promise handler. I've added a TODO one. This showed up because of my promise rejection code. 

**Benefits**

- When using the Project Files Pane you can switch between files that are big and small and it will cancel the previous load and start loading the new file you clicked.
- When using a large file (lego 9x18) you will be able to edit within the codemirror pane and it will restart the entire rendering instead of waiting for the entire renderer to finish before using your new code input.

### CODE REVIEW

File: src/lang/KclSingleton.ts [hunk 1]
```diff
@@ -232,6 +233,12 @@ export class KclManager {
   async executeAst(args: ExecuteArgs = {}): Promise<void> {
     if (this.isExecuting) {
       this.executeIsStale = args
+      // Previous execution will be rejected and finish, execution will be marked as stale
+      // new executeAst will start. This will reject previous AST calls while the user is typing
```
Comment (Irev-Dev) [inline src/lang/KclSingleton.ts:237]: Definitely a nit, because I'm talking about a comment, but "AST calls" doesn't seem like the right terminology. Maybe I misunderstood but would "will reject pending engine commands" make sense?

File: src/lang/KclSingleton.ts [hunk 2]
```diff
@@ -245,44 +252,47 @@ export class KclManager {
     // Make sure we clear before starting again. End session will do this.
     this.engineCommandManager?.endSession()
     await this.ensureWasmInit()
-    const { logs, errors, programMemory } = await executeAst({
+    const { logs, errors, programMemory, isInterrupted } = await executeAst({
```
Comment (Irev-Dev) [inline src/lang/KclSingleton.ts:255]: 👌

File: src/lang/std/engineConnection.ts [hunk 3]
```diff
@@ -1936,7 +1936,14 @@ export class EngineCommandManager extends EventTarget {
...
       range: [0, 0],
-    }).then(([a]) => a)
+    })
+      .then(([a]) => a)
+      .catch((e) => {
+        // TODO: Previously was never caught, since adding rejectAllCommands how should we handle this
+        // error?
+        /*noop*/
+        return null
+      })
```
Comment (Irev-Dev) [inline src/lang/std/engineConnection.ts:1946]: I think swallowing rejected scene commands is probably fine.

maybe 
```ts
if (e.msg ===
        'Failed to wait for promise from engine: JsValue("Force interrupt, executionIsStale, new AST requested")') {
  // swallow error
} else {
  // console.errror?
}
```

**Review by Irev-Dev — APPROVED** [review]
Really nice work!

File: src/lang/KclSingleton.ts [hunk 4]
```diff
@@ -232,6 +233,12 @@ export class KclManager {
   async executeAst(args: ExecuteArgs = {}): Promise<void> {
     if (this.isExecuting) {
       this.executeIsStale = args
+      // Previous execution will be rejected and finish, execution will be marked as stale
+      // new executeAst will start. This will reject previous AST calls while the user is typing
+      // in the editor
+      this.engineCommandManager.rejectAllCommands(
+        `Force interrupt, executionIsStale, new AST requested`
```
Comment (jtran) [inline src/lang/KclSingleton.ts:240]: Can this be a constant so that it can be referenced when we detect whether execution was interrupted?

**Comment by jtran** [comment]
I still think there's a race condition, or at least one waiting to happen, where the Rust isn’t currently waiting on any promise in `pendingCommands`. It could be awaiting something else. Then rejecting everything in `pendingCommands` won’t stop the executor (because `pendingCommands` is empty).

Even if there's no such thing now, I don't think there's a guarantee that everything that the WebAssembly module could await is stored in `pendingCommands`.

This is why the idea came up of having `sendModelingCommandFromWasm()` check if `executeIsStale` is true, and if so, throw.

File: src/lang/std/engineConnection.ts (hunk 3 above)
Comment (Irev-Dev) [reply src/lang/std/engineConnection.ts:1946]: No this won't work sorry.
send scene commands have nothing to do with rust, that's why you rejecting all pending commands causes this error. Mabye we need to differentiate between scene and modeling commands in pending commands and then do `rejectAllModelingCommands` instead of all modeling commands.

**Comment by Irev-Dev** [comment]
> It could be awaiting something else

What else could it be awaiting? afaik `sendModelingCommandFromWasm` is the only way that rust interfaces with the typescript so I don't understand what else it could be awaiting on?

The `executeIsStale` ideas is to send that flag to rust yeah? how would that be done, like next time rust calls `sendModelingCommandFromWasm` we return the result plus the flag, in which case we're waiting on the next time rust makes the call, and I guess I don't see how that's much different from just rejecting all the promises?

I'm on board just don't understand the mechanics, and rejecting the promises is kinda a hack since it's just giving rust back something it doesn't expect as a way to stop the exectution.

File: src/lang/std/engineConnection.ts (hunk 3 above)
Comment (nadr0) [reply src/lang/std/engineConnection.ts:1946]: I think you are correct on this. I was able to break the `initPlanes` commands in some workflows when spamming file switching or if the connection to the engine refreshed.

Can I use that type to see if they fall into the `Models['ModelingCmd_type]` ?

File: src/lang/KclSingleton.ts (hunk 1 above)
Comment (nadr0) [reply src/lang/KclSingleton.ts:237]: The reason I mention `AST` at all is the code that ends up returning is this block

```
    const { logs, errors, programMemory } = await executeAst({
      ast: newAst,
      engineCommandManager: this.engineCommandManager,
      useFakeExecutor: true,
    })
```

You are correct that the pending engine commands will be rejected but higher up the only reason those commands are being created is the `executeAst` call.

I can update the comment.

**Comment by nadr0** [comment]
> This is why the idea came up of having sendModelingCommandFromWasm() check if executeIsStale is true, and if so, throw.

> in which case we're waiting on the next time rust makes the call

I am open to other ideas for rejecting the promises but I think that waiting for another call to `sendModelingCommandFromWasm` to come in to cancel feels like it could cause problems as well?

What if it cannot send another command but the executor is still running? What if it takes X amount of time longer than expected?

I think in any solution we come up with, I feel as if we should be able to trigger a line of code that says please stop everything right now.

**Comment by jtran** [comment]
tl;dr: I think the PR is great as is. With 3 additional lines, I think it can also be robust to future changes that could cause subtle race condition bugs.

> The `executeIsStale` ideas is to send that flag to rust yeah? how would that be done, like next time rust calls `sendModelingCommandFromWasm` we return the result plus the flag, in which case we're waiting on the next time rust makes the call, and I guess I don't see how that's much different from just rejecting all the promises?

To be clear, I'm not suggesting an alternative as much as an additional change.

The idea is to additionally modify `sendModelingCommandFromWasm()` in `engineConnection.ts` so that before it calls `this.sendCommand()`, it does something like this:

```typescript
if (kclSingleton.executeIsStale) {
  return Promise.reject(new Error(...))
}
```

In the case that rejecting promises doesn't work, yes, we would be waiting on the next time that Rust calls `sendModelingCommandFromWasm()`. This approach is different from rejecting all promises because it would work even if there were no outstanding `pendingCommands` promises at the time when we reject them.

I think doing both would be good because rejecting promises prevents waiting when a promise exists. But the `executeIsStale` check guarantees that we don't try to send more commands to the engine when there weren't commands already sent.

> I'm on board just don't understand the mechanics, and rejecting the promises is kinda a hack since it's just giving rust back something it doesn't expect as a way to stop the exectution.

I admit that this new idea is hacky too, in its own way.

I actually think the promise rejection approach is clever and fine because the Rust code already has a mechanism for handling rejected promises. It's just an `Err` of the `Result` enum. It must handle this case because there could always be a network I/O error.

> What else could it be awaiting? afaik `sendModelingCommandFromWasm` is the only way that rust interfaces with the typescript so I don't understand what else it could be awaiting on?

I admit that I don't think it awaits anything else right now. But there's no guarantee that such a change won't happen in the future, so it makes me a little nervous.

But if you're sure @Irev-Dev that we don't have any other way to do anything async and we're not likely to ever do it, we can ignore this. I just don't like creating landmines for our future selves.

> I feel as if we should be able to trigger a line of code that says please stop everything right now.

Ignoring WebWorkers, JavaScript is single threaded. So if we're in a click handler determining that the user switched files, that means Rust is blocked on an await.

That's why the promise rejection is a great approach, IMO. It's exactly a way to say "stop everything right now". The only potential problem is that we don't have a reference to the particular promise that Rust is awaiting, which is why I'm suggesting this additional change.

Does that make sense? Sorry if I'm worrying for nothing.

**Comment by nadr0** [comment]
> That's why the promise rejection is a great approach, IMO. It's exactly a way to say "stop everything right now". The only potential problem is that we don't have a reference to the particular promise that Rust is awaiting, which is why I'm suggesting this additional change.

I like this. Thanks for taking the time to explain your thought process. I agree that if we have both we are limiting the chances of a race condition or poor behavior. Your suggestion of both is great since either can trigger first and the expected out come should be the same. 

In this [comment](https://github.com/KittyCAD/modeling-app/pull/3703#discussion_r1735355307) above I'll write some detection logic to only stop the modeling commands from going through in both of these locations. We will want the scene commands to go through. 

I really appreciate the discussion and suggestions, thanks Jon! I'll work on this tomorrow.

File: src/lang/std/engineConnection.ts (hunk 3 above)
Comment (Irev-Dev) [reply src/lang/std/engineConnection.ts:1946]: hmm so to clarify the modeling commands and scene commands is only a differentiation that we using in the modeling app, from the engine's perspective everything is a modeling command (with the exception of like webrtc setup command etc). Point is you're not going to be able to use types.

The break down is basically modeling commands are commands made by the rust execution
and everything the modeling app sends directly (mostly camera stuff) is a scene command. I'm being lazy in just sending diff screenshots, but this is what I think we should do

image

The type will need this added too

image

Than update your rejectAllCommands to something like
```ts
rejectAllModelingCommands(rejectionMessage: string) {
    Object.values(this.pendingCommands).forEach(
      ({ reject, isSceneCommand }) => !isSceneCommand && reject(rejectionMessage)
    )
  }
```

Does that make sense?

**Comment by Irev-Dev** [comment]
```
if (kclSingleton.executeIsStale) {
  return Promise.reject(new Error(...))
}
```
makes sense to me.

File: src/lang/KclSingleton.ts (hunk 1 above)
Comment (Irev-Dev) [reply src/lang/KclSingleton.ts:237]: hmm okay, no you can leave it. I just don't know what an "AST call" is, would "AST execution" work?

File: src/lang/std/engineConnection.ts (hunk 3 above)
Comment (nadr0) [reply src/lang/std/engineConnection.ts:1946]: Yup this makes sense, thanks for the detailed outline.

**Review by jtran — APPROVED** [review]
Nice!

**Comment by Irev-Dev** [comment]
image
btw this to me seems like it's most likely a circular dependency issue. looking at it atm.

File: src/lang/std/engineConnection.ts [hunk 5]
```diff
@@ -16,6 +16,8 @@ import { useModelingContext } from 'hooks/useModelingContext'
 import { exportMake } from 'lib/exportMake'
 import toast from 'react-hot-toast'
 import { SettingsViaQueryString } from 'lib/settings/settingsTypes'
+import { kclManager } from 'lib/singletons'
```
Comment (Irev-Dev) [inline src/lang/std/engineConnection.ts:19]: I suspect this is the problem, importing from the singletons module into one of the singletons.

Comment (Irev-Dev) [reply src/lang/std/engineConnection.ts]: Okay I've push ab13c25345b204cdefac2277e73fe3c78aa7aabd, hopefully build-test-web is happy now

**Review by Irev-Dev — APPROVED** [review]
Yay, it passed, going to enable auto merge for the test of the tests.

File: src/lang/std/engineConnection.ts (hunk 5 above)
Comment (nadr0) [reply src/lang/std/engineConnection.ts]: Thanks for catching this and fixing it. I appreciate it. I'll take a look into the import pattern you used.
